File: apps/windows/AreaMatrix/Features/Recovery/MissingFileRecoveryCoreBridge.cs
using System;
using System.Threading;
using System.Threading.Tasks;

namespace AreaMatrix.Features.Recovery;

public interface IMissingFileRecoveryCoreBridge
{
    Task<MissingFileRecoveryState> GetMissingFileStateAsync(
        string repoPath,
        long fileId,
        CancellationToken cancellationToken = default);

    Task<MissingFileRecoveryReport> RelinkMissingFileAsync(
        string repoPath,
        MissingFileRelinkRequest request,
        CancellationToken cancellationToken = default);

    Task<MissingFileRecoveryReport> RemoveMissingFileRecordAsync(
        string repoPath,
        MissingFileRemoveRecordRequest request,
        CancellationToken cancellationToken = default);
}

public interface IAreaMatrixMissingFileRecoveryCoreClient
{
    Task<CoreMissingFileState> GetMissingFileStateAsync(
        string repoPath,
        long fileId,
        CancellationToken cancellationToken = default);

    Task<CoreMissingFileRecoveryReport> RelinkMissingFileAsync(
        string repoPath,
        CoreMissingFileRelinkRequest request,
        CancellationToken cancellationToken = default);

    Task<CoreMissingFileRecoveryReport> RemoveMissingFileRecordAsync(
        string repoPath,
        CoreMissingFileRemoveRecordRequest request,
        CancellationToken cancellationToken = default);
}

public sealed class MissingFileRecoveryCoreBridge : IMissingFileRecoveryCoreBridge
{
    private readonly IAreaMatrixMissingFileRecoveryCoreClient coreClient;

    public MissingFileRecoveryCoreBridge(IAreaMatrixMissingFileRecoveryCoreClient coreClient)
    {
        this.coreClient = coreClient;
    }

    public async Task<MissingFileRecoveryState> GetMissingFileStateAsync(
        string repoPath,
        long fileId,
        CancellationToken cancellationToken = default)
    {
        CoreMissingFileState state = await coreClient
            .GetMissingFileStateAsync(repoPath, fileId, cancellationToken)
            .ConfigureAwait(false);
        return state.ToRecoveryState();
    }

    public async Task<MissingFileRecoveryReport> RelinkMissingFileAsync(
        string repoPath,
        MissingFileRelinkRequest request,
        CancellationToken cancellationToken = default)
    {
        CoreMissingFileRecoveryReport report = await coreClient
            .RelinkMissingFileAsync(
                repoPath,
                new CoreMissingFileRelinkRequest(request.FileId, request.NewPath, request.Confirmed),
                cancellationToken)
            .ConfigureAwait(false);
        return report.ToRecoveryReport();
    }

    public async Task<MissingFileRecoveryReport> RemoveMissingFileRecordAsync(
        string repoPath,
        MissingFileRemoveRecordRequest request,
        CancellationToken cancellationToken = default)
    {
        CoreMissingFileRecoveryReport report = await coreClient
            .RemoveMissingFileRecordAsync(
                repoPath,
                new CoreMissingFileRemoveRecordRequest(request.FileId, request.Confirmed),
                cancellationToken)
            .ConfigureAwait(false);
        return report.ToRecoveryReport();
    }
}

internal static class MissingFileRecoveryCoreMapping
{
    public static MissingFileRecoveryState ToRecoveryState(this CoreMissingFileState state)
    {
        return new MissingFileRecoveryState(
            state.FileId,
            state.RelativePath,
            state.LastKnownPath,
            state.LastSeenAt,
            ParseReason(state.Reason),
            state.ExpectedHashSha256,
            state.CanLocate,
            state.CanTryAgain,
            state.CanRemoveRecord,
            state.RemoveRecordRequiresConfirmation,
            state.CanRunRescan,
            state.RescanDisabledReason);
    }

    public static MissingFileRecoveryReport ToRecoveryReport(this CoreMissingFileRecoveryReport report)
    {
        return new MissingFileRecoveryReport(
            report.FileId,
            ParseStatus(report.Status),
            report.PreviousPath,
            report.CurrentPath,
            report.HashMatched,
            report.RecordRemoved,
            report.FileDeleted,
            report.ChangeLogAction,
            report.Message);
    }

    private static MissingFileReason ParseReason(string value)
    {
        return value switch
        {
            "PathMissing" => MissingFileReason.PathMissing,
            "PermissionDenied" => MissingFileReason.PermissionDenied,
            "CloudPlaceholder" => MissingFileReason.CloudPlaceholder,
            "ExternalVolumeDisconnected" => MissingFileReason.ExternalVolumeDisconnected,
            "Unknown" => MissingFileReason.Unknown,
            _ => throw ConfigError($"Unsupported missing-file reason `{value}`.")
        };
    }

    private static MissingFileRecoveryStatus ParseStatus(string value)
    {
        return value switch
        {
            "Missing" => MissingFileRecoveryStatus.Missing,
            "Present" => MissingFileRecoveryStatus.Present,
            "Relinked" => MissingFileRecoveryStatus.Relinked,
            "HashMismatch" => MissingFileRecoveryStatus.HashMismatch,
            "RecordRemoved" => MissingFileRecoveryStatus.RecordRemoved,
            "Blocked" => MissingFileRecoveryStatus.Blocked,
            _ => throw ConfigError($"Unsupported missing-file recovery status `{value}`.")
        };
    }

    private static MissingFileRecoveryCoreException ConfigError(string message)
    {
        return new MissingFileRecoveryCoreException(MissingFileRecoveryErrorKind.Config, message);
    }
}

public sealed class MissingFileRecoveryCoreException : Exception
{
    public MissingFileRecoveryCoreException(MissingFileRecoveryErrorKind kind, string message)
        : base(message)
    {
        Kind = kind;
    }

    public MissingFileRecoveryErrorKind Kind { get; }
}

File: apps/ios/AreaMatrix/Features/Recovery/MissingFileRecoveryCoreBridge.swift
import Foundation

protocol MissingFileRecoveryCoreBridge: Sendable {
    func getMissingFileState(repoPath: String, fileID: Int64) async throws -> MissingFileRecoveryState
    func relinkMissingFile(
        repoPath: String,
        request: MissingFileRelinkRequest
    ) async throws -> MissingFileRecoveryReport
    func removeMissingFileRecord(
        repoPath: String,
        request: MissingFileRemoveRecordRequest
    ) async throws -> MissingFileRecoveryReport
}

enum MissingFileReason: String, Equatable, Sendable {
    case pathMissing = "PathMissing"
    case permissionDenied = "PermissionDenied"
    case cloudPlaceholder = "CloudPlaceholder"
    case externalVolumeDisconnected = "ExternalVolumeDisconnected"
    case unknown = "Unknown"

    var displayText: String {
        switch self {
        case .pathMissing:
            "Path missing"
        case .permissionDenied:
            "Permission denied"
        case .cloudPlaceholder:
            "Cloud placeholder"
        case .externalVolumeDisconnected:
            "External volume disconnected"
        case .unknown:
            "Unknown"
        }
    }
}

enum MissingFileRecoveryStatus: String, Equatable, Sendable {
    case missing = "Missing"
    case present = "Present"
    case relinked = "Relinked"
    case hashMismatch = "HashMismatch"
    case recordRemoved = "RecordRemoved"
    case blocked = "Blocked"
}

struct MissingFileRecoveryState: Equatable, Sendable {
    var fileID: Int64
    var relativePath: String
    var lastKnownPath: String?
    var lastSeenAt: Int64?
    var reason: MissingFileReason
    var expectedHashSha256: String?
    var canLocate: Bool
    var canTryAgain: Bool
    var canRemoveRecord: Bool
    var removeRecordRequiresConfirmation: Bool
    var canRunRescan: Bool
    var rescanDisabledReason: String?

    var fileText: String {
        relativePath.isEmpty ? "No last known path is available." : relativePath
    }

    var lastKnownLocationText: String {
        guard let lastKnownPath, !lastKnownPath.isEmpty else {
            return "No last known path is available."
        }
        return lastKnownPath
    }

    var lastSeenText: String {
        guard let lastSeenAt, lastSeenAt > 0 else {
            return "Unknown"
        }
        return Date(timeIntervalSince1970: TimeInterval(lastSeenAt))
            .formatted(date: .abbreviated, time: .shortened)
    }

    var hashRequirementText: String {
        if expectedHashSha256?.isEmpty == false {
            return "Selected file must match the missing record hash."
        }
        return "Selected file will be checked by Core before relinking."
    }
}

struct MissingFileRelinkRequest: Equatable, Sendable {
    var fileID: Int64
    var newPath: String
    var confirmed: Bool
}

struct MissingFileRemoveRecordRequest: Equatable, Sendable {
    var fileID: Int64
    var confirmed: Bool
}

struct MissingFileRecoveryReport: Equatable, Sendable {
    var fileID: Int64
    var status: MissingFileRecoveryStatus
    var previousPath: String?
    var currentPath: String?
    var hashMatched: Bool
    var recordRemoved: Bool
    var fileDeleted: Bool
    var changeLogAction: String?
    var message: String?

    var displayMessage: String {
        if let message, !message.isEmpty {
            return message
        }
        switch status {
        case .relinked:
            return "File relinked."
        case .hashMismatch:
            return "Selected file does not match the missing record."
        case .recordRemoved:
            return "AreaMatrix record removed. No user file was deleted."
        case .present:
            return "File is available again."
        case .blocked:
            return "Recovery is blocked."
        case .missing:
            return "File is still missing."
        }
    }
}

enum MissingFileRecoveryError: Error, Equatable, Sendable {
    case fileNotFound(String)
    case database(String)
    case permissionDenied(String)
    case unavailable(String)

    var message: String {
        switch self {
        case .fileNotFound:
            "This missing file record is no longer available."
        case let .database(message):
            message.isEmpty ? "Repository metadata could not be updated." : message
        case .permissionDenied:
            "AreaMatrix does not have permission to complete this recovery action."
        case let .unavailable(message):
            message.isEmpty ? "Missing file recovery is unavailable." : message
        }
    }

    var recovery: String {
        switch self {
        case .fileNotFound:
            "Return to the file list and refresh."
        case .database:
            "Try again after the repository database is available."
        case .permissionDenied:
            "Check file and repository permissions, then try again."
        case .unavailable:
            "Try again."
        }
    }

    static func map(_ error: Error) -> MissingFileRecoveryError {
        if let recoveryError = error as? MissingFileRecoveryError {
            return recoveryError
        }
        if let detailError = error as? MobileFileDetailError {
            switch detailError {
            case let .fileNotFound(message):
                return .fileNotFound(message)
            case let .database(message):
                return .database(message)
            case let .permissionDenied(message):
                return .permissionDenied(message)
            case let .unavailable(message):
                return .unavailable(message)
            }
        }
        return .unavailable(error.localizedDescription)
    }
}

extension LiveMobileRepositoryCoreBridge: MissingFileRecoveryCoreBridge {
    func getMissingFileState(repoPath: String, fileID: Int64) async throws -> MissingFileRecoveryState {
        try await Task.detached(priority: .userInitiated) {
            try MissingFileRecoveryCoreFFIClient().getMissingFileState(repoPath: repoPath, fileID: fileID)
        }.value
    }

    func relinkMissingFile(
        repoPath: String,
        request: MissingFileRelinkRequest
    ) async throws -> MissingFileRecoveryReport {
        try await Task.detached(priority: .userInitiated) {
            try MissingFileRecoveryCoreFFIClient().relinkMissingFile(repoPath: repoPath, request: request)
        }.value
    }

    func removeMissingFileRecord(
        repoPath: String,
        request: MissingFileRemoveRecordRequest
    ) async throws -> MissingFileRecoveryReport {
        try await Task.detached(priority: .userInitiated) {
            try MissingFileRecoveryCoreFFIClient().removeMissingFileRecord(repoPath: repoPath, request: request)
        }.value
    }
}
